### PR TITLE
Update top level provider documentation for BackendV2

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

As a follow-on to #5885 this commit updates the top level
qiskit.providers module documentation for ``BackendV2``. This primarily
consists of two things, updating the writing a provider section to use
``BackendV2`` instead of ``BackendV1`` and adding a migration section on user
facing changes between ``BackendV1`` and ``BackendV2``.

### Details and comments

### CODE REVIEW

File: qiskit/providers/__init__.py [hunk 1]
```diff
@@ -118,22 +118,22 @@
...
+:mod:`qiskit.algorithms`, to get results from executing circuits on devices in a standard
+fashion regardless of how the backend is implemented. At a high level the basic
 steps for writing a provider are:
 
  * Implement a :class:`~qiskit.providers.ProviderV1` subclass that handles
    access to the backend(s).
- * Implement a :class:`~qiskit.providers.BackendV1` subclass and its
-   :meth:`~~qiskit.providers.BackendV1.run` method.
+ * Implement a :class:`~qiskit.providers.BackendV2` subclass and its
+   :meth:`~~qiskit.providers.BackendV2run` method.
```
Comment (jakelishman) [inline qiskit/providers/__init__.py:136]: missing `.` (I can't seem to suggest changes in single-commit view.)

File: qiskit/providers/__init__.py [hunk 2]
```diff
@@ -118,22 +118,22 @@
...
 
    * Add any custom gates for the backend's basis to the session
      :class:`~qiskit.circuit.EquivalenceLibrary` instance.
```
Comment (jakelishman) [inline qiskit/providers/__init__.py:139]: Is the "custom gates" bullet meant to be an indentation level higher than the others?

File: qiskit/providers/__init__.py [hunk 3]
```diff
@@ -240,21 +263,11 @@ def run(circuits, **kwargs):
...
-It's worth noting that this interface will likely evolve over time as the
-compiler grows and improves, thus requiring new information from the backends.
-This is a key reason why the providers interfaces are versioned to enable
-making these changes in a controlled manner. The details here will change and
-reflect the latest version of this interface.
+This is handled with the :class:`~qiskit.transpiler.Target` class which defines
+a model of a backend for the transpiler. A backend object will need to return
+a :class:`~qiskit.transpiler.Target` object from the :attr:`~qiskit.providers.BackendV2.target`
+attribute which the :func:`~qiskit.compiler.transpile` function will use as
+it's model of a backend target for compilation.
```
Comment (jakelishman) [inline qiskit/providers/__init__.py:270]: its model

File: qiskit/providers/__init__.py [hunk 4]
```diff
@@ -286,6 +299,28 @@ def _define(self):
 identify the gate in the basis set for the transpiler. If there is a conflict
 the transpiler will not know which gate to use.
 
+2. Add the custom gate to the target for your backend. This can be done with
+the :class:`~qiskit.transpiler.Target` class's
+:meth:`~qiskit.transpiler.Target.add_instruction` method. You'll need to
+add an instance of ``SYGate`` and it's parameters to the target so the transpiler
+knows it exists. For example, assuming this is part of your
+:class:~qiskit.providers.BackendV2` implementation for your backend::
```
Comment (jakelishman) [inline qiskit/providers/__init__.py:307]: This doesn't display as an actual ordered list if you wanted it to (continuation lines need indenting to match the "A" of "Add"), and the next list item is also numbered "2".

File: qiskit/providers/__init__.py [hunk 5]
```diff
@@ -286,6 +299,28 @@ def _define(self):
 identify the gate in the basis set for the transpiler. If there is a conflict
 the transpiler will not know which gate to use.
 
+2. Add the custom gate to the target for your backend. This can be done with
+the :class:`~qiskit.transpiler.Target` class's
+:meth:`~qiskit.transpiler.Target.add_instruction` method. You'll need to
```
Comment (jakelishman) [inline qiskit/providers/__init__.py:304]: It doesn't matter, but in the future if you prefer, you can also just write

> The can be done with the :meth:\`.Target.add_instruction\` method.

and it'll render `Target.add_instruction()`.

File: qiskit/providers/__init__.py [hunk 6]
```diff
@@ -184,37 +184,60 @@ def backends(self, name=None, **kwargs):
...
-            'coupling_map': [[0, 1], [1, 2], [2, 3], [3, 4]],
-            'simulator': False,
-            'local': False,
-            'conditional': True,
-            'open_pulse': False,
-            'memory': True,
-            'max_shots': 8196,
-            'gates': []
-        }
+        def __init(self):
```
Comment (jakelishman) [inline qiskit/providers/__init__.py:206]: `__init__`

File: qiskit/providers/__init__.py [hunk 7]
```diff
@@ -184,37 +184,60 @@ def backends(self, name=None, **kwargs):
...
+            cx_props = {edge: None for edge in [(0, 1), (1, 2), (2, 3), (3, 4)]}
+            self._target.add_instruction(CXGate(), cx_props)
+            meas_props = {(qubit,): None for qubit in range(5)}
+            self._target.add_instruction(Measure(), meas_props)
+            id_props = {(qubit,): None for qubit in range(5)}
+            self._target.add_instruction(IGate(), id_props)
+
+            # Set option validators
+            self.options.set_validator("shots", (1, 4096))
+            self.options.set_validator("measure", bool)
```
Comment (jakelishman) [inline qiskit/providers/__init__.py:232]:
```python
KeyError: "Field 'measure' is not present in this options object"
```

File: qiskit/providers/__init__.py [hunk 8]
```diff
@@ -286,6 +299,28 @@ def _define(self):
 identify the gate in the basis set for the transpiler. If there is a conflict
 the transpiler will not know which gate to use.
 
+2. Add the custom gate to the target for your backend. This can be done with
+the :class:`~qiskit.transpiler.Target` class's
+:meth:`~qiskit.transpiler.Target.add_instruction` method. You'll need to
+add an instance of ``SYGate`` and it's parameters to the target so the transpiler
```
Comment (jakelishman) [inline qiskit/providers/__init__.py:305]: its parameters

File: qiskit/providers/__init__.py (hunk 4 above)
Comment (jakelishman) [inline qiskit/providers/__init__.py:307]: missing `` ` `` in the Sphinx directive

File: qiskit/providers/__init__.py [hunk 9]
```diff
@@ -392,21 +427,32 @@ def run(self, circuits. **kwargs):
...
 object with all the default values for all the options a backend supports. For
 example, if the backend supports only supports ``shots`` the
-:obj:`~qiskit.providers.BackendV1._default_options` method would look like::
+:obj:`~qiskit.providers.BackendV2._default_options` method would look like::
 
     @classmethod
     def _default_options(cls):
         return Options(shots=1024)
 
+Tou can also set validators on an :class:`~qiskit.providers.Options` object to
```
Comment (jakelishman) [inline qiskit/providers/__init__.py:440]: You

File: qiskit/providers/__init__.py [hunk 10]
```diff
@@ -500,6 +546,110 @@ def result(self):
...
+======================
+
+The :obj:`~BackendV2` class re-defined user access for most properties of a
+backend to make them work with native Qiskit data structures and have flatter
+access patterns. However this means when using a provider that upgrades
+from :obj:`~BackendV1` to :obj:`~BackendV2` existing access patterns will need
+to be adjusted. It is expected for existing providers to deprecate the old
+access where possible to provide a graceful migration, but eventually users
+will need to adjust code. The biggest change to adapt to in :obj:`~BackendV2` is
+that most of the information accesible about a backend is contained in it's
```
Comment (jakelishman) [inline qiskit/providers/__init__.py:563]: contained in its

File: qiskit/providers/__init__.py [hunk 11]
```diff
@@ -500,6 +546,110 @@ def result(self):
...
+:class:`~qiskit.transpiler.Target` object and the backend's attributes often query
+it's :attr:`~qiskit.providers.BackendV2.target`
```
Comment (jakelishman) [inline qiskit/providers/__init__.py:565]: its

File: qiskit/providers/__init__.py [hunk 12]
```diff
@@ -500,6 +546,110 @@ def result(self):
...
+attribute to return information, however in many cases the attributes only provide
+a subset of information the target can contain. For example, ``backend.coupling_map``
+returns a :class:`~qiskit.transpiler.CouplingMap` constructed from the
+:class:`~qiskit.transpiler.Target` accesible in the
+:attr:`~qiskit.providers.BackendV2.target` attribute, however the target may contain
+instructions that operate on more than two qubits (which can't be represented in a
+:class:`~qiskit.transpiler.CouplingMap`) or has instructions that only operate on
+a subset of qubits (or two qubit links for a two qubit instruction) which won't be
+detailed in the full coupling map returned by
+:attr:`~qiskit.providers.backend_v2.coupling_map`. So depending on your use case
```
Comment (jakelishman) [inline qiskit/providers/__init__.py:575]: This crossref looks funky - it's not linked, and its role is `attr` but there's no class in it.

File: qiskit/providers/__init__.py [hunk 13]
```diff
@@ -500,6 +546,110 @@ def result(self):
...
+   * - :obj:`~BackendV1`
+     - :obj:`~BackendV2`
+     - Notes
+   * - ``backend.configuration().n_qubits``
+     - ``backend.num_qubits``
+     -
+   * - ``backend.configuration().coupling_map``
+     - ``backend.coupling_map``
+     - The return from :obj:`~BackendV2` is a :class:`~qiskit.transpiler.CouplingMap` object.
+       while in :obj:`~BackendV1` it is jsut an edge list. Also this is just a view of
```
Comment (jakelishman) [inline qiskit/providers/__init__.py:594]: jsut -> just [an edge list]

File: qiskit/providers/__init__.py [hunk 14]
```diff
@@ -500,6 +546,110 @@ def result(self):
...
+       the information contained in ``backend.target`` which may only be a subset of the
+       information contained in :class:`~qiskit.transpiler.Target` object.
```
Comment (jakelishman) [inline qiskit/providers/__init__.py:596]: I just noticed it now - nothing is generating the `Target` doc stubs, so all the `Target` crossrefs fail.  I imagine it needs an `autoclass` or an `autosummary` in the `transpiler` docstring (probably in the other PR).

**Review by jakelishman — COMMENTED** [review]
Just a quick read through. I don't know enough about what's happened in the main `BackendV2` PR to properly review the content, but the bits I do know looked correct to me.

File: qiskit/providers/__init__.py (hunk 2 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py:139]: It's supposed to be a sub-bullet on the implement the backendV2 bullet point (as it's part of implementing the backend class).

File: qiskit/providers/__init__.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Fixed in the current revision (just forced pushed as part of a rebase too)

File: qiskit/providers/__init__.py (hunk 3 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Fixed in the current revision (just forced pushed as part of a rebase too)

File: qiskit/providers/__init__.py (hunk 4 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Fixed in the current revision (just forced pushed as part of a rebase too)

File: qiskit/providers/__init__.py (hunk 5 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Fixed in the current revision (just forced pushed as part of a rebase too)

File: qiskit/providers/__init__.py (hunk 6 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Fixed in the current revision (just forced pushed as part of a rebase too)

File: qiskit/providers/__init__.py (hunk 7 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Oops, yeah it should be `"memory"` not `"measure"`. Fixed in the current revision (just forced pushed as part of a rebase too)

File: qiskit/providers/__init__.py (hunk 8 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Fixed in the current revision (just forced pushed as part of a rebase too)

File: qiskit/providers/__init__.py (hunk 4 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Fixed in the current revision (just forced pushed as part of a rebase too)

File: qiskit/providers/__init__.py (hunk 9 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Fixed in the current revision (just forced pushed as part of a rebase too)

File: qiskit/providers/__init__.py (hunk 10 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Fixed in the current revision (just forced pushed as part of a rebase too)

File: qiskit/providers/__init__.py (hunk 11 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Fixed in the current revision (just forced pushed as part of a rebase too)

File: qiskit/providers/__init__.py (hunk 12 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Fixed in the current revision (just forced pushed as part of a rebase too)

File: qiskit/providers/__init__.py (hunk 13 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Fixed in the current revision (just forced pushed as part of a rebase too)

File: qiskit/providers/__init__.py (hunk 14 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py:595]: Yeah good catch. It's missing from the autodoc in `transpiler/__init__.py` in #5885. It's re-exported from `__init__.py` but for some reason I didn't add it to an autosummary directive. I'll update that now over there and then rebase this after

Comment (mtreinish) [reply qiskit/providers/__init__.py:595]: Ok, I updated #5885 to include `Target` in the docs and rebased this now on top of that so the crossrefs should work now.

**Comment by mtreinish** [comment]
Ok, now that #5885 has merged I've rebased this on the merged version and it's no longer blocked

File: qiskit/providers/__init__.py [hunk 15]
```diff
@@ -240,122 +263,133 @@ def run(circuits, **kwargs):
 --------------------
 
 The key piece of the backend object is how it describes itself to the compiler.
```
Comment (1ucian0) [inline qiskit/providers/__init__.py:265]:
```suggestion
The key piece of the :class:`~qiskit.providers.Backend` object is how it describes itself to the compiler.
```

**Comment by 1ucian0** [comment]
The following sentence sounds repetitive:
```
For a simple example of a provider the
`qiskit-aqt-provider <https://github.com/qiskit-community/qiskit-aqt-provider>`__
package is worth using as an example.
```
I suggest:
```
For a simple example of a provider, see the
`qiskit-aqt-provider <https://github.com/qiskit-community/qiskit-aqt-provider>`__
package.
```

File: qiskit/providers/__init__.py [hunk 16]
```diff
@@ -118,22 +118,22 @@
...
 
  * Implement a :class:`~qiskit.providers.ProviderV1` subclass that handles
    access to the backend(s).
- * Implement a :class:`~qiskit.providers.BackendV1` subclass and its
-   :meth:`~~qiskit.providers.BackendV1.run` method.
+ * Implement a :class:`~qiskit.providers.BackendV2` subclass and its
+   :meth:`~~qiskit.providers.BackendV2.run` method.
 
    * Add any custom gates for the backend's basis to the session
      :class:`~qiskit.circuit.EquivalenceLibrary` instance.
```
Comment (1ucian0) [inline qiskit/providers/__init__.py:139]: Is this spacing intentional?


Screenshot 2021-12-01 at 21 44 55

```suggestion
   :meth:`~~qiskit.providers.BackendV2.run` method.
   * Add any custom gates for the backend's basis to the session
     :class:`~qiskit.circuit.EquivalenceLibrary` instance.
```

File: qiskit/providers/__init__.py [hunk 17]
```diff
@@ -240,122 +263,133 @@ def run(circuits, **kwargs):
...
-know how to work in the SY basis. To solve this the ``SYGate`` class would
-need to be added to Qiskit and
-:class:`~qiskit.quantum_info.OneQubitEulerDecomposer` updated to support
-decomposing to the ``SYGate``. Longer term that is likely a better direction
-for custom basis gates and contributing the definitions and support in the
-transpiler will ensure that it continues to be well supported by Qiskit
-moving forward.
+   (:mod:`qiskit.circuit.library`) you can integrate support for this into your
+   provider. The basic method for doing this is first to define a
+   :class:`~qiskit.circuit.Gate` class for all the custom gates in the basis
```
Comment (1ucian0) [inline qiskit/providers/__init__.py:278]:
```suggestion
   :class:`~qiskit.circuit.Gate` subclass for all the custom gates in the basis
```

File: qiskit/providers/__init__.py [hunk 18]
```diff
@@ -240,122 +263,133 @@ def run(circuits, **kwargs):
...
+           (1,): InstructionProperties(duration=2.1e-6, error=0.0001)
+           (2,): InstructionProperties(duration=2.5e-6, error=0.0003)
+           (3,): InstructionProperties(duration=2.2e-6, error=0.0004)
+       }
+       self.target.add_instruction(SYGate(), sy_props)
+
+   The keys in ``sy_props`` define the qubits on the backend ``SYGate`` can be
+   used on, and the values define the properties of ``SYGate`` on that qubit.
+   For multiqubit gates the tuple keys contain all qubit combinations the gate
+   works on (order is significant ``(0, 1)`` is different from ``(1, 0)``).
```
Comment (1ucian0) [inline qiskit/providers/__init__.py:321]:
```suggestion
   works on (order is significant, ie ``(0, 1)`` is different from ``(1, 0)``).
```

File: qiskit/providers/__init__.py [hunk 19]
```diff
@@ -240,122 +263,133 @@ def run(circuits, **kwargs):
...
+
+3. After you've defined the custom gates to use for the Backend's basis set
```
Comment (1ucian0) [inline qiskit/providers/__init__.py:323]:
```suggestion
3. After you've defined the custom gates to use for the backend's basis set
```

File: qiskit/providers/__init__.py [hunk 20]
```diff
@@ -500,6 +545,110 @@ def result(self):
...
+   * - :obj:`~BackendV1`
+     - :obj:`~BackendV2`
+     - Notes
+   * - ``backend.configuration().n_qubits``
+     - ``backend.num_qubits``
+     -
+   * - ``backend.configuration().coupling_map``
+     - ``backend.coupling_map``
+     - The return from :obj:`~BackendV2` is a :class:`~qiskit.transpiler.CouplingMap` object.
+       while in :obj:`~BackendV1` it is just an edge list. Also this is just a view of
```
Comment (1ucian0) [inline qiskit/providers/__init__.py:593]:
```suggestion
       while in :obj:`~BackendV1` it is an edge list. Also this is just a view of
```

File: qiskit/providers/__init__.py (hunk 16 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py:139]: It is, the custom gates thing is a sub-bullet of building a backend (as it's an optionally a part of building a backend class) see where @jakelishman asked the same thing: https://github.com/Qiskit/qiskit-terra/pull/7323#discussion_r759069763

**Comment by mtreinish** [comment]
> The following sentence sounds repetitive:
> 
> ```
> For a simple example of a provider the
> `qiskit-aqt-provider <https://github.com/qiskit-community/qiskit-aqt-provider>`__
> package is worth using as an example.
> ```
> 
> I suggest:
> 
> ```
> For a simple example of a provider, see the
> `qiskit-aqt-provider <https://github.com/qiskit-community/qiskit-aqt-provider>`__
> package.
> ```

Done in: https://github.com/Qiskit/qiskit-terra/pull/7323/commits/254270d5e01dce620f907912dd94d5560a5fc066 I also updated the link to the repo's current home

**Review by 1ucian0 — APPROVED** [review]
Thanks!

**Review by 1ucian0 — APPROVED** [review]

**Pull request merged**